### PR TITLE
Extend KFServing component with autoscaling and server mode

### PR DESCRIPTION
**Pull request by Tomcli**
Extend KFServing component with the following features:
- Add autoscaling annotations
- Add server mode to accept external requests for model serving.
- minor code refactor

/assign @animeshsingh

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2315)

### CODE REVIEW

File: components/kubeflow/kfserving/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 FROM python:3.6-slim
 
-RUN pip install kubernetes==9.0.0 kfserving==0.1.1 requests==2.22.0
+RUN pip install kubernetes==9.0.0 kfserving==0.1.1 requests==2.22.0 Flask==1.1.1 flask-cors==3.0.8
```
Comment (Ark-kun) [inline components/kubeflow/kfserving/Dockerfile]: Nit. It's a good habit to use `pip3`, since some images have `pip` pointing to python2.

Comment (Tomcli) [reply components/kubeflow/kfserving/Dockerfile]: Thanks. Done.

File: components/kubeflow/kfserving/component.yaml [hunk 2]
```diff
@@ -10,11 +10,13 @@ inputs:
   - {name: Framework,                       type: String, default: 'tensorflow', description: 'Machine Learning Framework for Model Serving.'}
   - {name: default_custom_model_spec,       type: String, default: '{}', description: 'Custom runtime default custom model container spec.'}
   - {name: canary_custom_model_spec,        type: String, default: '{}', description: 'Custom runtime canary custom model container spec.'}
+  - {name: Autoscaling Target,              type: String, default: '0',        description: 'Autoscaling Target Number'}
+  - {name: KFServing Deployer API,          type: String, default: '', description: 'Custom runtime canary custom model container spec.'}
```
Comment (animeshsingh) [inline components/kubeflow/kfserving/component.yaml]: description doesn't match

File: components/kubeflow/kfserving/component.yaml [hunk 3]
```diff
@@ -10,11 +10,13 @@ inputs:
   - {name: Framework,                       type: String, default: 'tensorflow', description: 'Machine Learning Framework for Model Serving.'}
   - {name: default_custom_model_spec,       type: String, default: '{}', description: 'Custom runtime default custom model container spec.'}
```
Comment (animeshsingh) [inline components/kubeflow/kfserving/component.yaml]: Should be named like other

File: components/kubeflow/kfserving/src/kfservingdeployer.py [hunk 4]
```diff
@@ -122,8 +104,74 @@ def kfserving_deployment(metadata, default_model_spec, canary_model_spec=None, c
...
+            canary_model_traffic=canary_model_traffic,
+            namespace=namespace,
+            framework=framework,
+            default_custom_model_spec=default_custom_model_spec,
+            canary_custom_model_spec=canary_custom_model_spec,
+            autoscaling_target=autoscaling_target
+        )
+    print(model_status)
+    try:
+        print(model_status['status']['url'] + ' is the knative domain header. $CLUSTER_IP:31380 is the endpoint to post predictions to the model.')
```
Comment (animeshsingh) [inline components/kubeflow/kfserving/src/kfservingdeployer.py]: why 31830 is hardcoded?

File: components/kubeflow/kfserving/src/kfservingdeployer.py [hunk 5]
```diff
@@ -122,8 +104,74 @@ def kfserving_deployment(metadata, default_model_spec, canary_model_spec=None, c
...
+        print('Sample test curl below: ')
+        print('curl -X GET -H "Host: ' + url.sub('', model_status['status']['url']) + '" $CLUSTER_IP:31380')
+    except:
```
Comment (animeshsingh) [inline components/kubeflow/kfserving/src/kfservingdeployer.py]: Same here?

File: components/kubeflow/kfserving/component.yaml (hunk 2 above)
Comment (animeshsingh) [reply components/kubeflow/kfserving/component.yaml]: Also change this to kfserving_endpoint

File: components/kubeflow/kfserving/sample-pipeline.py [hunk 6]
```diff
@@ -28,7 +28,11 @@ def kfservingPipeline(
     canary_model_uri='gs://kfserving-samples/models/tensorflow/flowers',
```
Comment (animeshsingh) [inline components/kubeflow/kfserving/sample-pipeline.py]: I would suggest removing this file. We will come up with a sample later in samples repo imho

Comment (animeshsingh) [inline components/kubeflow/kfserving/sample-pipeline.py]: if we keep, shall be flowers-2 for canary

File: components/kubeflow/kfserving/component.yaml (hunk 2 above)
Comment (Tomcli) [reply components/kubeflow/kfserving/component.yaml]: Done

File: components/kubeflow/kfserving/component.yaml (hunk 3 above)
Comment (Tomcli) [reply components/kubeflow/kfserving/component.yaml]: Done

File: components/kubeflow/kfserving/sample-pipeline.py (hunk 6 above)
Comment (Tomcli) [reply components/kubeflow/kfserving/sample-pipeline.py]: Done

File: components/kubeflow/kfserving/src/kfservingdeployer.py (hunk 4 above)
Comment (Tomcli) [reply components/kubeflow/kfserving/src/kfservingdeployer.py]: The default istio ingress gateway is hosted with nodePort 31380. I updated this section with some bash command to get the LoadBalancer IP to make it more clarify.

File: components/kubeflow/kfserving/component.yaml [hunk 7]
```diff
@@ -8,13 +8,15 @@ inputs:
   - {name: Canary Model Traffic Percentage, type: String, default: '0',        description: 'Optional Traffic to be sent to default model'}
   - {name: Namespace,                       type: String, default: 'kubeflow', description: 'Kubernetes namespace where the KFServing service is deployed.'}
   - {name: Framework,                       type: String, default: 'tensorflow', description: 'Machine Learning Framework for Model Serving.'}
-  - {name: default_custom_model_spec,       type: String, default: '{}', description: 'Custom runtime default custom model container spec.'}
-  - {name: canary_custom_model_spec,        type: String, default: '{}', description: 'Custom runtime canary custom model container spec.'}
+  - {name: Default Custom Model Spec,       type: String, default: '{}', description: 'Custom runtime default custom model container spec.'}
```
Comment (Ark-kun) [inline components/kubeflow/kfserving/component.yaml]: JFYI: If you use `type: JsonObject`, then the argument to this input will be automatically JSON-serialized.
e.g. `kfserve_op(..., spec={'model_version': 1.7}), ...`.

**Pull request merged**